src/librustc_front/visit.rs

```diff
@@ -309,91 +302,63 @@ pub fn walk_variant<'v, V: Visitor<'v>>(visitor: &mut V,
     match variant.node.kind {
         TupleVariantKind(ref variant_arguments) => {
             for variant_argument in variant_arguments {
-                visitor.visit_ty(&*variant_argument.ty)
+                visitor.visit_ty(&variant_argument.ty)
             }
         }
         StructVariantKind(ref struct_definition) => {
-            visitor.visit_struct_def(&**struct_definition,
+            visitor.visit_struct_def(struct_definition,
                                      variant.node.name,
                                      generics,
                                      variant.node.id)
         }
     }
-    match variant.node.disr_expr {
-        Some(ref expr) => visitor.visit_expr(&**expr),
-        None => ()
-    }
-    for attr in &variant.node.attrs {
-        visitor.visit_attribute(attr);
-    }
-}
-
-pub fn skip_ty<'v, V: Visitor<'v>>(_: &mut V, _: &'v Ty) {
-    // Empty!
-}
-
-pub fn walk_ty_opt<'v, V: Visitor<'v>>(visitor: &mut V, optional_type: &'v Option<P<Ty>>) {
-    match *optional_type {
-        Some(ref ty) => visitor.visit_ty(&**ty),
-        None => ()
-    }
+    walk_list!(visitor, visit_expr, &variant.node.disr_expr);
+    walk_list!(visitor, visit_attribute, &variant.node.attrs);
 }
 
 pub fn walk_ty<'v, V: Visitor<'v>>(visitor: &mut V, typ: &'v Ty) {
     match typ.node {
         TyVec(ref ty) | TyParen(ref ty) => {
-            visitor.visit_ty(&**ty)
+            visitor.visit_ty(ty)
         }
         TyPtr(ref mutable_type) => {
-            visitor.visit_ty(&*mutable_type.ty)
+            visitor.visit_ty(&mutable_type.ty)
         }
-        TyRptr(ref lifetime, ref mutable_type) => {
-            visitor.visit_opt_lifetime_ref(typ.span, lifetime);
-            visitor.visit_ty(&*mutable_type.ty)
+        TyRptr(ref opt_lifetime, ref mutable_type) => {
+            walk_list!(visitor, visit_lifetime, opt_lifetime);
+            visitor.visit_ty(&mutable_type.ty)
         }
         TyTup(ref tuple_element_types) => {
-            for tuple_element_type in tuple_element_types {
-                visitor.visit_ty(&**tuple_element_type)
-            }
+            walk_list!(visitor, visit_ty, tuple_element_types);
         }
         TyBareFn(ref function_declaration) => {
-            for argument in &function_declaration.decl.inputs {
-                visitor.visit_ty(&*argument.ty)
-            }
-            walk_fn_ret_ty(visitor, &function_declaration.decl.output);
-            walk_lifetime_decls_helper(visitor, &function_declaration.lifetimes);
+            walk_fn_decl(visitor, &function_declaration.decl);
+            walk_list!(visitor, visit_lifetime_def, &function_declaration.lifetimes);
         }
         TyPath(ref maybe_qself, ref path) => {
-            if let Some(ref qself) = *maybe_qself {
+            for qself in maybe_qself {
```

Comment:
Not sure I like this change - `if let` seems more appropriate than `for` here - why the change?